File: cpp/prob010.cpp
#include <iostream>
#include <chrono>
#include "euler.h"

#define PROB_NUM    10

uint64_t compute()
{
    std::vector<uint32_t> primes = list_prime(2000000);
    std::vector<uint64_t> primes_int(primes.begin(), primes.end());
    return sum(primes_int);
}


int test_answer(std::string prob_answer) {
    std::string answer;
    if(!get_answer(PROB_NUM, &answer)) {
        std::cout << "file not found" << std::endl;
        return 1;
    }

    assert(answer.compare(prob_answer) == 0);
    return 0;
}

int main(int argc, char const *argv[])
{
    auto t1 = std::chrono::high_resolution_clock::now();
    int64_t result = compute();
    auto t2 = std::chrono::high_resolution_clock::now();

    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(t2 - t1).count();
    std::cout << result << "\tin " << duration << " us\n";

    // do testing
    test_answer(std::to_string(result));
    return 0;
}

File: rust/src/bin/prob010.rs
extern crate euler;

fn compute() -> u64 {
    euler::list_prime(2000000).iter()
        .fold(0u64, |acc, x| acc + *x as u64)
}

fn main() {
    let now = std::time::Instant::now();
    let result = compute();
    println!("{:?}\tin {}ms", result, now.elapsed().as_millis());
}

#[cfg(test)]
mod tests {
    use super::*;
    const PROB_NUM: usize = 10;
    
    #[test]
    fn test_prob010() {
        let result = compute().to_string();
        euler::test_answer(PROB_NUM, result);
    }
}